### PR TITLE
Integrate memkind

### PR DESCRIPTION
**Pull request by michalbiesek**
- add memkind allocator variant
- required memkind version 1.10

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/redis/120)

### CODE REVIEW

File: src/config.c [hunk 1]
```diff
@@ -1998,12 +1998,12 @@ static void numericConfigRewrite(typeData data, const char *name, struct rewrite
 }
 
 static int isValidActiveDefrag(int val, char **err) {
-#ifndef HAVE_DEFRAG
+#if !defined(HAVE_DEFRAG) && !defined(HAVE_DEFRAG_MEMKIND)
     if (val) {
         *err = "Active defragmentation cannot be enabled: it "
                "requires a Redis server compiled with a modified Jemalloc "
                "like the one shipped by default with the Redis source "
-               "distribution";
```
Comment (jschmieg) [inline src/config.c:2006]: , like the one shipped by default with the Redis source distribution, or memkind.

Comment (michalbiesek) [reply src/config.c:2006]: Done

File: tests/unit/memefficiency.tcl [hunk 2]
```diff
@@ -37,7 +37,6 @@ start_server {tags {"memefficiency"}} {
 }
 
 start_server {tags {"defrag"}} {
-    if {[string match {*jemalloc*} [s mem_allocator]]} {
         test "Active defrag" {
```
Comment (kasiawasiuta) [inline tests/unit/memefficiency.tcl:41]: Please fix indentation here, everything else looks good :)

Comment (michalbiesek) [reply tests/unit/memefficiency.tcl:41]: Done by extend the match pattern -  ci/jenkins/redis and ci/jenkins/redis_memkind configuration job should be updated @kasiawasiuta

**Review by jschmieg — APPROVED** [review]
> Reviewed 3 of 7 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/120)* status: all files reviewed, 1 unresolved discussion (waiting on @kasiawasiuta and @KFilipek)

**Review by KFilipek — CHANGES_REQUESTED** [review]
> Reviewed 3 of 7 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/120)* status: all files reviewed, 2 unresolved discussions (waiting on @kasiawasiuta and @michalbiesek)

___
*[src/defrag.c, line 82 at r2](https://reviewable.io/reviews/pmem/redis/120#-M-9erjP5X2i_3PSos-o:-M-9erjQ3JxXzXsMFYws:b-uzliw2) ([raw file](https://github.com/pmem/redis/blob/8ba56b74f1b1ba4d9af2ee2c3969be9a2136bcfb/src/defrag.c#L82)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> void* activeDefragAlloc(void *ptr) {
>   void* new = memkind_defrag_reallocate(NULL, ptr);
>   if (!new) server.stat_active_defrag_misses++;
>   return new;
> }
> ```
</details>

In other part of this file indentation is 4 spaces instead of 2.

**Review by KFilipek — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/120)* status: all files reviewed, 3 unresolved discussions (waiting on @kasiawasiuta and @michalbiesek)

___
*[src/defrag.c, line 81 at r2](https://reviewable.io/reviews/pmem/redis/120#-M-9imrj2stwJ6bYnch2:-M-9imrk5QvnlsTrKzpb:b-eddfsa) ([raw file](https://github.com/pmem/redis/blob/8ba56b74f1b1ba4d9af2ee2c3969be9a2136bcfb/src/defrag.c#L81)):*
> ```C
> #ifdef HAVE_DEFRAG_MEMKIND
> void* activeDefragAlloc(void *ptr) {
>   void* new = memkind_defrag_reallocate(NULL, ptr);
> ```

I'm not sure that using "new" name is good practice, but it's not used as a keyword in C.

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/120)* status: 6 of 7 files reviewed, 3 unresolved discussions (waiting on @jschmieg, @kasiawasiuta, and @KFilipek)

___
*[src/defrag.c, line 81 at r2](https://reviewable.io/reviews/pmem/redis/120#-M-9imrj2stwJ6bYnch2:-M-9qjQS939rymGV7Byq:b-896fix) ([raw file](https://github.com/pmem/redis/blob/8ba56b74f1b1ba4d9af2ee2c3969be9a2136bcfb/src/defrag.c#L81)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

I'm not sure that using "new" name is good practice, but it's not used as a keyword in C.
</blockquote></details>

Done.
___
*[src/defrag.c, line 82 at r2](https://reviewable.io/reviews/pmem/redis/120#-M-9erjP5X2i_3PSos-o:-M-9qjDJDDOzXBlwmtQr:b-896fix) ([raw file](https://github.com/pmem/redis/blob/8ba56b74f1b1ba4d9af2ee2c3969be9a2136bcfb/src/defrag.c#L82)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

> ```
> void* activeDefragAlloc(void *ptr) {
>   void* new = memkind_defrag_reallocate(NULL, ptr);
>   if (!new) server.stat_active_defrag_misses++;
>   return new;
> }
> ```
In other part of this file indentation is 4 spaces instead of 2.
</blockquote></details>

Done.

**Review by KFilipek — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/120)* status: all files reviewed, 1 unresolved discussion (waiting on @kasiawasiuta)

**Pull request merged**